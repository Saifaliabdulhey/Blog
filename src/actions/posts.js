import database from '../firebase/firebase'

// ADD POSTS
export const addPost = (post) => ({
  type: "ADD_POST",
  post
});
export const startAddPosts = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      description = ''
    } = postData;

    const post = { description, title };

    return database.ref(`users/${uid}/posts`).push(post).then((ref) => {
      dispatch(addPost({
        id: ref.key,
        ...post
      }))
    })
  }
}


//REMOVE POST
export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`).remove().then(() => {
      dispatch(removePost({ id }));
    })
  }
}


//EDIT_POST
export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
})

export const startEditpost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`)
      .update(updates)
      .then(() => dispatch(editPost(id, updates)))
  }
}

// SET_POSTS 
export const setPosts = (posts) => ({
  type: "SET_POSTS",
  posts
})
export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      const posts = [];
      snapshot.forEach((childSnapshot) => {
        posts.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setPosts(posts));
    })
  }
}
