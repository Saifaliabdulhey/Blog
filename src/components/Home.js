import React from 'react';
import PostListFilters from './PostListFilters';
import PostList from './PostList';

function Home(props) {
  return (
    <div>
      <PostListFilters />
      <PostList /> 
    </div>

  )
}

export default Home
