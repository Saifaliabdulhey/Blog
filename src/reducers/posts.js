const postsReducerDefaultState = [];
  
const postsReducer = (state = postsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_POST":
      return [...state, action.post];
      case 'EDIT_POST':
        return state.map((post) => {
          if(post.id === action.if) {
            return {
              ...post, 
              ...action.updates
            };
          }else {
            return post;
          }
        })
      case 'REMOVE_POST':
        return state.filter(({ id }) => id !== action.id)
      case 'SET_POSTS':
        return action.posts;
    default:
      return state;
  }
};


export default postsReducer;