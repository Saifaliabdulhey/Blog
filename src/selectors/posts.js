const getVisiblePosts = (posts, text) => {
    return posts.filter((post) => {
        const textMatch = post.title
            .toLowerCase()
            .includes(text.toLowerCase());
        return textMatch
    })
}


export default getVisiblePosts;