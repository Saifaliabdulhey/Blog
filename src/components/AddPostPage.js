import React, { Component } from 'react';
import PostForm from './PostForm';
import { connect } from 'react-redux';
import { startAddPosts } from '../actions/posts'

export class AddPostPage extends Component {

    onSubmit = (post) => {
        this.props.startAddPosts(post);
        this.props.history.push("/home")
    }
    render() {
        return (
            <div>
                <div className="container">
                    <h1>Add Post</h1>
                </div>
                <div className="container">
                    <PostForm  onSubmit={this.onSubmit}/>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddPosts: (post) => dispatch(startAddPosts(post))
})

export default connect(undefined, mapDispatchToProps)(AddPostPage);
