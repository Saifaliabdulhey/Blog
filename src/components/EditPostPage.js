import React, { Component, setState } from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startRemovePost, startEditpost } from '../actions/posts';
import { Link } from 'react-router-dom'

export class EditPostPage extends Component {
    onSubmit = (post) => {
        this.props.startEditpost(this.props.post.id, post);
        this.props.history.push('/home');
    }
    onRemove = () => {
        this.props.startRemovePost({ id: this.props.post.id });
        this.props.history.push('/home');
    }

    render() {
        let path = window.location.origin;
        const id = this.props.match.params.id;
        return (
            <div>
                <div className="container">
                    <h1>Edit Post</h1>
                </div>
                <div className="container">
                    <PostForm
                        post={this.props.post}
                        onSubmit={this.onSubmit} />
                    <div className="container btn-container">
                        <button  onClick={this.onRemove}>Remove Post</button>
                    </div>
                    <Link className="read_link" to={`/read/${id}`}>
                    <h4><span>Post is Readable at:</span>{`${path}/read/${id}`}</h4>
                    </Link>
                    
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    post: state.posts.find((post) => post.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
    startEditpost: (id, post) => dispatch(startEditpost(id, post)),
    startRemovePost: (data) => dispatch(startRemovePost(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);