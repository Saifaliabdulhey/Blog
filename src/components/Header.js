import React  from 'react'
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'
import { startLogout } from '../actions/auth'





export const Header = ({ startLogout }) => (

    <header className="header">
        <div className="container">
            <div className="header_content">
                <Link to="/home"><h1>My Blog</h1></Link>
                <button className="btn" onClick={startLogout}>logout</button>
            </div>
        </div>
    </header>

)

const mapStateToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})


export default connect(undefined, mapStateToProps)(Header)