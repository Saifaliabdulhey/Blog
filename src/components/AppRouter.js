import React from 'react';
import { Router, Route, Switch} from 'react-router-dom';
import AddPostPage from './AddPostPage';
import Home from './Home';
import createHistory from 'history/createBrowserHistory';
import EditPostPage from './EditPostPage';
import LoginPage from './LoginPage';
import Header from './Header';
import Read from './Read'

export const history = createHistory();

function AppRouter() {
    return (
        <Router history={history}>
            <Header />
            <Switch>
                <Route path="/" exact component={LoginPage} />
                <Route path="/home" exact component={Home} />
                <Route path="/read/:id" exact component={Read} />
                <Route path="/edit/:id" component={EditPostPage} />
                <Route path="/add" exact component={AddPostPage} />
            </Switch>
        </Router>
    )
}

export default AppRouter