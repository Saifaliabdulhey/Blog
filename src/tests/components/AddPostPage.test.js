import React from 'react';
import { shallow } from 'enzyme';
import { AddPostPage } from '../../components/AddPostPage'
import posts from '../fixture/posts'

let startAddPosts, history, wrapper;

beforeEach(() => {
    startAddPosts = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddPostPage startAddPosts={startAddPosts} history={history} />);
});

test('should render AddPostsPage correctly', () => {
    expect(wrapper).toMatchSnapshot();
})


// test('should handle onSubmit', () => {
//     wrapper.find('PostForm').prop('onSubmit')(posts[1]);
//     expect(history.push).toHaveBeenLastCalledWith('/');
//     expect(startAddPosts).toHaveBeenLastCalledWith(posts[1]);
// })