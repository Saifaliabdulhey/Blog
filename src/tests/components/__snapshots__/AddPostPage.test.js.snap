// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddPostsPage correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <div
      className="container"
    >
      <h1>
        Add Post
      </h1>
    </div>
    <div
      className="container"
    >
      <componentName
        onSubmit={[Function]}
      />
    </div>
  </div>,
  "nodes": Array [
    <div>
      <div
        className="container"
      >
        <h1>
          Add Post
        </h1>
      </div>
      <div
        className="container"
      >
        <componentName
          onSubmit={[Function]}
        />
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_didScheduleRenderPhaseUpdate": false,
    "_dispatcher": Object {
      "readContext": [Function],
      "useCallback": [Function],
      "useContext": [Function],
      "useDebugValue": [Function],
      "useDeferredValue": [Function],
      "useEffect": [Function],
      "useId": [Function],
      "useImperativeHandle": [Function],
      "useInsertionEffect": [Function],
      "useLayoutEffect": [Function],
      "useMemo": [Function],
      "useReducer": [Function],
      "useRef": [Function],
      "useResponder": [Function],
      "useState": [Function],
      "useSyncExternalStore": [Function],
      "useTransition": [Function],
    },
    "_element": <AddPostPage
      history={
        Object {
          "push": [MockFunction],
        }
      }
      startAddPosts={[MockFunction]}
    />,
    "_firstWorkInProgressHook": null,
    "_forcedUpdate": false,
    "_idCounter": 0,
    "_instance": AddPostPage {
      "context": Object {},
      "onSubmit": [Function],
      "props": Object {
        "history": Object {
          "push": [MockFunction],
        },
        "startAddPosts": [MockFunction],
      },
      "refs": Object {},
      "state": null,
      "updater": Updater {
        "_callbacks": Array [],
        "_renderer": [Circular],
      },
    },
    "_isReRender": false,
    "_newState": null,
    "_numberOfReRenders": 0,
    "_renderPhaseUpdates": null,
    "_rendered": <div>
      <div
        className="container"
      >
        <h1>
          Add Post
        </h1>
      </div>
      <div
        className="container"
      >
        <componentName
          onSubmit={[Function]}
        />
      </div>
    </div>,
    "_rendering": false,
    "_updater": Updater {
      "_callbacks": Array [],
      "_renderer": [Circular],
    },
    "_workInProgressHook": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AddPostPage
    history={
      Object {
        "push": [MockFunction],
      }
    }
    startAddPosts={[MockFunction]}
  />,
}
`;
